### PR TITLE
Feature 106

### PR DESCRIPTION
**Pull request by Bjerkk**
Feature #106 is the functionality to prompt user when pictogram search takes a long time

### CODE REVIEW

**Comment by KVISDAOWNER** [comment]
@aau-giraf/sw609f19 you have been assigned this. Here's your checklist. I marked @Tricky12321 as the reviewer, but you might change it as you like.
Please use the following checklist. Feel free to check it off as you go along.
**Code Design**
- [x] Is the code in the right place? (Both in terms of folder structure and class structure)
- [x] Could this code, to the best of your knowledge, have reused existing code?
- [x] Is the code over-engineered? Examples of over-engineering: Implemented behavior for future problems
- [x] Does it introduce functionality that is not necessary for solving the problem?

**Code Readability and Maintainability**
- [x] Are names meaningful and self-documenting?
- [x] Is it understandable, by reading the code, what it does?
- [x] Is the code simple and readable? i.e. it contains no “hack”, or obscure solution.
- [x] Is it understandable, by reading the tests, what they do?
- [x] Do the tests cover sensible cases? Both happy paths and exception paths?
- [x] Do the tests cover the full functionality? i.e., will the tests fail if some of the requested functionality is missing?
- [x] Are error messages in the code understandable?
- [x] Are confusing sections of code covered by comments?

**Functionality**
For this part, open the app, and test:
- [x] Are the features that the code claims to implement, actually implemented?
- [x] Does it look like the code has bugs?
- [x] Are all new screens reachable through in-app navigation?

**Points for Discussion and Reflection**
- [x] Are UI messages, that the user might meet, actually helpful?
- [x] Are there errors that are bad practice? Like, hard-coding a test database connection?
- [x] Is there well written code in this? Remember to tell the developer!

**Comment by KVISDAOWNER** [comment]
@aau-giraf/sw613f19 you have been assigned this. Here's your checklist. I marked @MrTasty  as the reviewer, but you might change it as you like.
Please use the following checklist. Feel free to check it off as you go along.
**Code Design**
- [x] Is the code in the right place? (Both in terms of folder structure and class structure)
- [x] Could this code, to the best of your knowledge, have reused existing code?
- [x] Is the code over-engineered? Examples of over-engineering: Implemented behavior for future problems
- [x] Does it introduce functionality that is not necessary for solving the problem?

**Code Readability and Maintainability**
- [x] Are names meaningful and self-documenting?
- [x] Is it understandable, by reading the code, what it does?
- [x] Is the code simple and readable? i.e. it contains no “hack”, or obscure solution.
- [x] Is it understandable, by reading the tests, what they do?
- [x] Do the tests cover sensible cases? Both happy paths and exception paths?
- [x] Do the tests cover the full functionality? i.e., will the tests fail if some of the requested functionality is missing?
- [x] Are error messages in the code understandable?
- [x] Are confusing sections of code covered by comments?

**Functionality**
For this part, open the app, and test:
- [x] Are the features that the code claims to implement, actually implemented?
- [x] Does it look like the code has bugs?
- [x] Are all new screens reachable through in-app navigation?

**Points for Discussion and Reflection**
- [x] Are UI messages, that the user might meet, actually helpful?
- [x] Are there errors that are bad practice? Like, hard-coding a test database connection?
- [x] Is there well written code in this? Remember to tell the developer!

File: lib/providers/api/user_api.dart [hunk 1]
```diff
@@ -0,0 +1,122 @@
+import 'package:api_client/http/http.dart';
+import 'package:api_client/models/giraf_user_model.dart';
+import 'package:api_client/models/settings_model.dart';
+import 'package:api_client/models/username_model.dart';
+import 'package:flutter/material.dart';
+import 'package:rxdart/rxdart.dart';
+
+/// User endpoints
+class UserApi {
```
Comment (boisenme) [inline lib/providers/api/user_api.dart]: This file should not be included as we now have a package for the API

File: lib/blocs/pictogram_bloc.dart [hunk 2]
```diff
@@ -42,23 +42,31 @@ class PictogramBloc extends BlocBase {
...
-
     _pictograms.add(null);
+    List<PictogramModel> _resultPlaceholder;
 
-    _timer = Timer(Duration(milliseconds: _milliseconds), () {
+    _debounceTimer = Timer(Duration(milliseconds: _debounceTime), () {
+      //Timer for sending an error if getting pictogram results takes too long
+      Timer(Duration(milliseconds: _timeoutTime), () {
+        if (_resultPlaceholder == null || _resultPlaceholder.isEmpty) {
+          _pictograms.addError(null);
```
Comment (boisenme) [inline lib/blocs/pictogram_bloc.dart]: Er det muligt at adde en fejlbesked her som så kan bruges i screenen istedet for null? Dette vil også gør det muligt at have flere forskellige fejlbeskeder.

File: lib/screens/pictogram_search_screen.dart [hunk 3]
```diff
@@ -38,19 +38,34 @@ class PictogramSearch extends StatelessWidget {
...
+                              .toList(),
+                        );
+
+                      }
+                      else if(snapshot.hasError){
+                        return const InkWell(
+                          key: Key('timeoutWidget'),
+                          child: Padding(
+                            padding: EdgeInsets.symmetric(horizontal: 20),
+                            child: Text('Søgningen gav ingen resultater. '
```
Comment (boisenme) [inline lib/screens/pictogram_search_screen.dart]: Her kan fejlbeskeden så bruges fra erroren

**Review by boisenme — DISMISSED** [review]
Works very well

Picture of when no connection
image

Small comments otherwise fine

**Review by msimon16 — DISMISSED** [review]

**Review by msimon16 — DISMISSED** [review]

**Review by msimon16 — APPROVED** [review]

**Review by andpaacoke — APPROVED** [review]
This has been PO approved 👍

**Review by AlsenDev — APPROVED** [review]

**Pull request merged**